service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,209 @@
+                giftCertificate.setTags(newTags);
+                id = dao.insert(giftCertificate);
+                GiftCertificate certificateWithAllTags = dao.findById(id).get();
+                certificateWithAllTags.setTags(SetUtils.union(newTags, existingTags));
+                dao.update(certificateWithAllTags);
+            } else {
+                id = dao.insert(giftCertificate);
+            }
+        } else {
+            throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ERROR, giftCertificate.toString());
+        }
+        return id;
+    }
+
+    @Override
+    public boolean delete(String id) {
+        try {
+            GiftCertificate giftCertificate = dao.findById(Long.parseLong(id)).orElseThrow(() ->
+                    new ResourceNotFoundException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                            ErrorAttribute.RESOURCE_NOT_FOUND_ERROR, id));
+            if (!CollectionUtils.isEmpty(giftCertificate.getTags())) {
+                dao.disconnectAllTags(giftCertificate);
+            }
+            return dao.delete(giftCertificate.getId());
+        } catch (NumberFormatException e) {
+            throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ID_ERROR, id);
+        }
+    }
+
+    @Override
+    public boolean update(String id, GiftCertificate newCertificate) {
+        try {
+            GiftCertificate oldCertificate = dao.findById(Long.parseLong(id)).orElseThrow(() ->
+                    new ResourceNotFoundException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                            ErrorAttribute.RESOURCE_NOT_FOUND_ERROR, id));
+
+            if (updateCertificateFields(oldCertificate, newCertificate)) {
+                oldCertificate.setLastUpdateDate(LocalDateTime.now());
+                if (!CollectionUtils.isEmpty(oldCertificate.getTags())) {
+                    List<Tag> newTags = (List<Tag>) CollectionUtils.removeAll(oldCertificate.getTags(),
+                            CollectionUtils.intersection(tagService.findAll(0, 0), oldCertificate.getTags()));
+                    newTags.forEach(tagService::insert);
+                }
+                dao.update(oldCertificate);
+            } else {
+                throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                        ErrorAttribute.INVALID_GIFT_CERTIFICATE_ERROR, newCertificate.toString());
+            }
+        } catch (NumberFormatException e) {
+            throw new InvalidFieldException(ErrorAttribute.GIFT_CERTIFICATE_ERROR_CODE,
+                    ErrorAttribute.INVALID_GIFT_CERTIFICATE_ID_ERROR, id);
+        }
+        return true;
+    }
+
+    @Override
+    public void disconnectTagById(String tagId) {
```

Comment:
Remember our discussion about the order and certificate.. But don't understand exactly what is going on here.